build/xtask/src/lsp.rs

```diff
@@ -0,0 +1,390 @@
+        build_override_command
+            .extend(dependencies.keys().map(|p| format!("-p{p}")));
+        build_override_command.push("--target-dir".to_owned());
+        build_override_command.push("target/rust-analyzer".to_owned());
+        build_override_command
+            .push(format!("--features={}", features.join(",")));
+
+        let mut out = LspConfig {
+            features,
+            target: target.unwrap().to_string(),
+            extra_env: build_cfg.env,
+            hash: "".to_owned(),
+            build_override_command,
+            app: app_name.clone().to_owned(),
+            task: task_name.to_owned(),
+        };
+
+        let mut s = DefaultHasher::new();
+        out.hash(&mut s);
+        out.hash = format!("{:x}", s.finish());
+        out.hash.truncate(8);
+
+        return Some(out);
+    }
+    None
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+fn inner(file: &PathBuf, clients: &[LspClient]) -> Result<LspConfig> {
+    if !file.is_file() {
+        bail!("input must be a file");
+    }
+    let mut dir = file
+        .parent()
+        .ok_or_else(|| anyhow!("could not find parent of {file:?}"))?;
+
+    let mut cargo = loop {
+        if let Ok(f) = std::fs::File::open(dir.join("Cargo.toml")) {
+            break f;
+        }
+        dir = dir
+            .parent()
+            .ok_or_else(|| anyhow!("reached root of filesystem"))?;
+    };
+
+    // Read the Cargo.toml to get the package name
+    let mut cargo_text = Vec::new();
+    cargo
+        .read_to_end(&mut cargo_text)
+        .context("failed to read Cargo.toml")?;
+    let cargo_toml: Manifest = toml::from_slice(&cargo_text)?;
+    let package_name = cargo_toml.package.name;
+
+    let metadata = cargo_metadata::MetadataCommand::new()
+        .no_deps()
+        .features(cargo_metadata::CargoOpt::AllFeatures)
+        .exec()
+        .context("failed to run cargo metadata")?;
+
+    let root = metadata.workspace_root.clone();
+    let packages = PackageGraph::new(metadata);
+
+    // First, see if we have a matching client in the list of clients provided
+    // to the CLI.  This minimizes the number of unique `rust-analyzer`
+    // configurations running simultaneously, and speeds up initial attach time.
+    for c in clients {
+        // TODO: we parse the PackageConfig multiple times here, which may be
+        // slow (but probably not slower than `cargo metadata` above)
+        let file = root.join(&c.toml);
+        let app_cfg = PackageConfig::new(&file, false, false)
+            .context(format!("could not open {file:?}"))?;
+        if let Some(out) =
+            check_task(&package_name, &c.task, &c.toml, &app_cfg, &packages)
+        {
+            return Ok(out);
+        }
+    }
+
+    let preferred_apps = [
+        "app/gimlet/rev-c.toml",
+        "app/sidecar/rev-b.toml",
+        "app/psc/rev-b.toml",
+    ];
```

Comment:
Is there a way to tell from within nvim which app it picked? When I open `control-plane-agent` I'm seeing some errors that imply none of the `gimlet`/`sidecar`/`psc` features are active, and I think I expected `gimlet` to be active based on this list.

Reply:
`LspInfo` should show you what app + task is selected.